### PR TITLE
[ENH] Added '|' dunder method to MultiplexForecaster

### PR DESCRIPTION
**Pull request by miraep8**
#### Reference Issues/PRs

Implements part of #2459. **Note this PR is dependent on PR #2458**

#### What does this implement/fix? Explain your changes.

Have implemented a `|` dunder method for both `MultiplexForecaster`/`BaseForecaster`. Have also introduced new tests for this code.

Other things in this PR:
- added a small `_get_forecaster_names()` helper method to `_HeterogenousEnsembleForecaster` seemed useful to me!
- updated the example code in for using `MultiplexForecaster` so that it no longer imports from all.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

In addition to the dunder behaving as expected - would be great to get feedback on how I decided to do the unique name handling.

Essentially a use case I was imagining was that a user might want to do this:
``
multiplex_forecaster = AutoETS() | NaiveForecaster() | ThetaForecaster()
``
or potentially even this:
``
multiplex_forecaster = NaiveForecaster(method="last") | NaiveForecaster(method="mean") | NaiveForecaster(method="drift")
``

In these cases we would want to generate unique names for them.   Would appreciate feedback on whether the approach I took to do this seems reasonable. (mostly I tried to leverage the `super()` methods which already existed and were relevant)

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [DOC] or [BUG] indicating whether the PR topic is related to enhancement, documentation or bug

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

First of all, I want to note that the "heterogeneous ensemble" estimators are a bit in flux currently, so feedback is appreciated.

Questions and/or change requests:

* I was intending to remove the `_HeterogeneousEnsembleForecaster` gradually, using only the `_HeterogenousMetaEstimator`. Do you think this is a good idea or not? It would avoid proliferation of the "intermediate" methods, like the `_get_forecaster_names` you added. Why not use `_get_estimator_names` directly?
* why did you decide to make an `_ensure_unique_names` method? This seems to duplicate `_get_estimator_tuples` which you can get from `_HeterogenousMetaEstimator` (in your case, with `clone_ests=True` arg). Am I missing something? Or did you miss the `_get_estimator_tuples` function?

File: sktime/forecasting/base/_base.py [hunk 1]
```diff
@@ -170,6 +170,27 @@ def __rmul__(self, other):
...
+        other: `sktime` forecaster or sktime MultiplexForecaster
+
+        Returns
+        -------
+        MultiplexForecaster object
+        """
+        from sktime.forecasting.compose import MultiplexForecaster
+
+        # generate a name for our forecaster based on class name:
+        self_name = type(self).__name__.lower()
```
Comment (fkiraly) [inline sktime/forecasting/base/_base.py:189]: I think the `MultiplexForecaster` shouhld take care of generating the name - I would just do `MultiplexForecaster(self)`

File: sktime/forecasting/base/_base.py [hunk 2]
```diff
@@ -170,6 +170,27 @@ def __rmul__(self, other):
...
+        # create wrap self in MultiplexForecaster, pass most of '|' logic to that class
+        multiplex_self = MultiplexForecaster([(self_name, self)])
+        return multiplex_self | other
```
Comment (fkiraly) [inline sktime/forecasting/base/_base.py:192]: we need to make a case distinction - you do the above if `other` is an instance of `BaseForecaster`, otherwise you need to return `NotImplemented` (see the other dunders)

File: sktime/forecasting/compose/_multiplexer.py [hunk 3]
```diff
@@ -102,6 +103,78 @@ def _check_selected_forecaster(self):
...
+            other_forecasters = self._ensure_unique_names(other.forecasters)
+            new_multiplex_forecaster = MultiplexForecaster(
+                self.forecasters + other_forecasters
+            )
+            return new_multiplex_forecaster
+        # If is anyother type of forecaster, simply add it to forecasters:
+        elif isinstance(other, BaseForecaster):
+            other_name = type(other).__name__.lower()
+            other_tuple = self._ensure_unique_names([(other_name, other)])
+            self.forecasters.append(other_tuple[0])
```
Comment (fkiraly) [inline sktime/forecasting/compose/_multiplexer.py:170]: here, I would construct a new `MultiplexForecaster`, instead of modifying the current one. Otherwise we get all kinds of weird side effects from partially fitted chains etc.

Comment (fkiraly) [reply sktime/forecasting/compose/_multiplexer.py:170]: in all similar places as well

**Comment by fkiraly** [comment]
Also, another note.

Overall, this is really nice work!

But, I think you seem to have probably not seen my advice where I said, that I have already been refactoring the `MultiplexForecaster` here: https://github.com/alan-turing-institute/sktime/pull/2458
and I have been using the "unique name" methods and done the integration with the new heterogeneous ensemble forecaster

I was also under the impression you would look at the `MultiplexTransformer`.

The two PR on the `MultiplexForecaster` are now, unfortunately, clashing.

Would you mind taking my PR https://github.com/alan-turing-institute/sktime/pull/2458 as a starting point, and try to reconcile it with the dunder methods you have added?

Sorry for the hassle, I should have been more clearer.

And, please let this not discourage you!
Your PR is one of the very highest quality pieces of work on `sktime` I've seen, I am serious.

File: sktime/forecasting/base/_base.py (hunk 2 above)
Comment (miraep8) [reply sktime/forecasting/base/_base.py:192]: So - two things I would appreciate some clarity on!  
1) I actually intententionally had the code return a `ValueError` if other isn't a `BaseForecaster` or a `MultiplexForecaster` instead of `NotImplemented`. (My impression is that [they plan to phase out NotImplemented](https://docs.python.org/3/library/constants.html#NotImplemented), so I decided to just use an error instead).  But given that would you prefer I stick with the `NotImplemented` all the same?   
2) is your issue that there are not checks (as there are - they are just wrapped up in the `MultiplexForecaster` `__or__`) or that you want them in 2 places?  I can understand the logic for the second as it makes it clearer what is going on at all stages, but does lead to code bloat a bit too..... Hence the clarification question :)

Comment (miraep8) [reply sktime/forecasting/base/_base.py:192]: Whoops!  Just kidding - I think I misread the python docs page!  Just went back and reread it!  I guess they are just depreciating the use of it in a boolean context not all together!

Comment (miraep8) [reply sktime/forecasting/base/_base.py:192]: In which case I am happy to stick with it since that is what the other dunders use!  Though personally I might find a `ValueError` more useful haha.

File: sktime/forecasting/compose/_multiplexer.py (hunk 3 above)
Comment (miraep8) [reply sktime/forecasting/compose/_multiplexer.py:170]: Sounds good will do!

File: sktime/forecasting/base/_base.py (hunk 1 above)
Comment (miraep8) [reply sktime/forecasting/base/_base.py:189]: ok!

**Comment by miraep8** [comment]
Hey!

Sorry! My bad, I had of course seen your PR #2458, it just hadn't initially seemed to me that we had too many clashes (at least in logic!) as believe the dunder method should work with your PR as well! (I tried my best not to touch any of the other functions in MultiplexForecaster so that hopefully we would avoid any direct conflicts!) I guess I also figured since you made your PR first it would probably get merged first and then I would handle all the conflicts on my end as needed!  I think I was perhaps unclear in my comments in #2459  - I had actually intended to first do the dunder for `MultiplexForecaster` to get familiar with how this works before implementing `MultiplexTransformer` to try to be consistent with how you guys have built these objects already!

As for the base classes - I am more than happy to switch over to using all the base classes in `_HeterogenousMetaEstimator` instead of `_HeterogenousMetaForecaster`.  The `_get_forecaster_names()` method was just a convenience to not have to type `multiplexer._get_estimator_names(multiplexer.forecasters)`, but I can definitely see the logic with getting away with `_HeterogenousMetaForecaster` entirely - at the moment it doesn't seem to add too much! I will do a commit with these changes later this evening!  I had actually somehow missed the `get_estimator_tuples` method, so in that refactor I will revisit it! 

Long story short - sorry about the clashes, I thought we were on the same page! Lets just chalk it up to an unfortunate miscommunication! I would certainly be happy to handle any merge conflicts that arise as a result! :)  Why don't we just leave this PR until after #2458 is merged and I will take care of it then.  

Thanks for the helpful feedback, I will make another commit handling some of your suggestions soon!

**Comment by fkiraly** [comment]
>Long story short - sorry about the clashes, I thought we were on the same page! Lets just chalk it up to an unfortunate miscommunication! I would certainly be happy to handle any merge conflicts that arise as a result! :)

My suggestion would be: create another branch, off your fork branch `multiplex_additions`. Into that branch, merge #2458. This will lead to (possibly horrific) conflicts, but you can develop off what is likely the end state of #2458.

Of course, #2458 could change based on review (incl comments that you leave there!), but I do not expect that to be severe.

**Comment by fkiraly** [comment]
feel free to make another draft PR, btw - that will let allow you to run the remote CI. Of course, you can also use `check_estimator` locally, that should also work.

**Comment by fkiraly** [comment]
generally, there are two advanced techniques relevant here if you are working on a major feature.
"feature stream" - a feature branch into which smaller PR are merged, then the entire feature branch is merged into dev/main
"chunk & chain" - multiple PR that depend on each other, and you start merging them in sequence (first to last)

If we had planned in advance to work on the same thing together, I would have gone for the first, since dunder & name/hierarchical seem independent. Now I'm suggesting the second - another example is #2496, #2500, #2502.

**Comment by fkiraly** [comment]
(win 3.8 failure is due to `pytest-timout` - I'll raise the limit in #2541 to 10min)

**Comment by miraep8** [comment]
Done! :) Fortunately as expected our only merge conflict ended up being the `[author]` line in multiplex.py, and otherwise all our changes were compatible! Thanks for the tips regarding best practices for the major feature - will be very useful going forward! I will go ahead and change the comments in this commit to reflect the fact that this PR is "Dependent" on your PR!

File: sktime/forecasting/base/_base.py (hunk 2 above)
Comment (fkiraly) [reply sktime/forecasting/base/_base.py]: I think the convention with dunders is returning `NotImplemented` if they are not implemented, and it is important to do so.

Why? There is a dunder parse order, let's say we have an expression `a * b * c`. If `a` does not implement `__mul__` for `b`, then python looks next whether `__rmul__` of `b` implements for `a`, which can be the case, and so on.

Had you raised an exception in `a.__mul__`, this would have broken in the first step of parsing. But, expected behaviour and implementation is that python needs to check whether one of the two objects implements it. The information is passed via `NotImplemented`, without raising an exception.

**Comment by fkiraly** [comment]
> I am curious if we actually needed to merge?

I'm surprised and a little confused. Didn't you do all that stuff in the `__init__` and with name unification? Surely that must be clashing on some level. Or so I thought.

But, maybe I overestimated the trouble, and you did a good job in encapsulating your changes even within the class. If that's the case (which it seems), kudos.

File: sktime/forecasting/base/_base.py (hunk 2 above)
Comment (miraep8) [reply sktime/forecasting/base/_base.py]: No worries - it is implemented like that in the new code! :)

**Comment by miraep8** [comment]
>  Didn't you do all that stuff in the __init__ and with name unification? Surely that must be clashing on some level. Or so I thought.

Hmm, I played around with the `__init__` in one of my first commits, but by the time I actually made the PR (several commits later) I got rid of all that because it was messy :) I can see how that could have been confusing though!

File: sktime/forecasting/base/_base.py (hunk 2 above)
Comment (miraep8) [reply sktime/forecasting/base/_base.py]: But this is helpful to know/makes sense

**Comment by fkiraly** [comment]
Hm, seems to work!

From my perspective, we could merge - but, due to the dependence on my own PR #2458, we need to wait for someone to review that.
You can´t approve since you are not a core dev, but you could help other reviewers by looking at it and pasting your remarks or change requests. Due to your contributions here, you are probably currently the other person who understands the `MultiplexForecaster` best.

**Review by fkiraly — APPROVED** [review]
Very nice contribution!

In the end, the merge was less problematic than I thought.

Will merge as soon as #2458 is merged.
